src/lib/grammar_ast.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+pub struct GrammarAST {
```

Comment:
The `GrammarAST` representation is all strings; the `Grammar` representation is (more-or-less) all integers. But you can't create a `Grammar` until you've read in the full input (i.e. the `GrammarAST`) because rules can make forward references.